src/t_hash.c

```diff
@@ -32,233 +32,119 @@
-        sdsSetAuxBit(embedded_field_sds, FIELD_SDS_AUX_BIT_ENTRY_HAS_VALUE_PTR, 1);
-        serverAssert(entryHasValuePtr(embedded_field_sds));
-    }
-    return (void *)embedded_field_sds;
-}
-
-/* The entry pointer is the field sds, but that's an implementation detail. */
-sds hashTypeEntryGetField(const hashTypeEntry *entry) {
-    return (sds)entry;
-}
-
-sds hashTypeEntryGetValue(const hashTypeEntry *entry) {
-    if (entryHasValuePtr(entry)) {
-        return *hashTypeEntryGetValueRef(entry);
-    } else {
-        /* Skip field content, field null terminator and value sds8 hdr. */
-        size_t offset = sdslen(entry) + 1 + sdsHdrSize(SDS_TYPE_8);
-        return (char *)entry + offset;
-    }
-}
-
-/* Returns the address of the entry allocation. */
-static void *hashTypeEntryAllocPtr(hashTypeEntry *entry) {
-    char *buf = sdsAllocPtr(entry);
-    if (entryHasValuePtr(entry)) {
-        buf -= sizeof(sds *);
-    }
-    return buf;
-}
-
-/* Frees previous value, takes ownership of new value, returns entry (may be
- * reallocated). */
-static hashTypeEntry *hashTypeEntryReplaceValue(hashTypeEntry *entry, sds value) {
-    sds field = (sds)entry;
-    size_t field_size = sdsHdrSize(sdsType(field)) + sdsalloc(field) + 1;
-    size_t value_len = sdslen(value);
-    size_t value_size = sdsReqSize(value_len, SDS_TYPE_8);
-    if (!entryHasValuePtr(entry)) {
-        /* Reuse the allocation if the new value fits and leaves no more than
-         * 25% unused space after replacing the value. */
-        char *alloc_ptr = sdsAllocPtr(entry);
-        size_t required_size = field_size + value_size;
-        size_t alloc_size;
-        if (required_size <= EMBED_VALUE_MAX_ALLOC_SIZE &&
-            required_size <= (alloc_size = hashTypeEntryMemUsage(entry)) &&
-            required_size >= alloc_size * 3 / 4) {
-            /* It fits in the allocation and leaves max 25% unused space. */
-            sdswrite(alloc_ptr + field_size, alloc_size - field_size, SDS_TYPE_8, value, value_len);
-            sdsfree(value);
-            return entry;
-        }
-        hashTypeEntry *new_entry = hashTypeCreateEntry(hashTypeEntryGetField(entry), value);
-        freeHashTypeEntry(entry);
-        return new_entry;
-    } else {
-        /* The value pointer is located before the embedded field. */
-        if (field_size + value_size <= EMBED_VALUE_MAX_ALLOC_SIZE) {
-            /* Convert to entry with embedded value. */
-            hashTypeEntry *new_entry = hashTypeCreateEntry(field, value);
-            freeHashTypeEntry(entry);
-            return new_entry;
-        } else {
-            /* Not embedded value. */
-            sds *value_ref = hashTypeEntryGetValueRef(entry);
-            sdsfree(*value_ref);
-            *value_ref = value;
-            return entry;
+void hashTypeFreeVolatileSet(robj *o) {
+    volatile_set *set = hashTypeGetVolatileSet(o);
+    if (set)
+        freeVolatileSet(set);
```

Comment:
nit. also free functions should work with null pointers, for consistency, such as `freeVolatileSet(hashTypeGetVolatileSet(o))`
```suggestion
    if (set) freeVolatileSet(set);
```

Reply:
> also free functions should work with null pointers

Fully agree. this is totally changes in the followup PRs so I feel there is no urgent need to change it here.